relay-system/src/monitor.rs

```diff
@@ -0,0 +1,191 @@
+use std::future::Future;
+use std::pin::Pin;
+use std::sync::atomic::{AtomicU64, AtomicU8, Ordering};
+use std::sync::Arc;
+use std::task::{Context, Poll};
+use tokio::time::{Duration, Instant};
+
+/// Minimum interval when utilization is recalculated.
+const UTILIZATION_UPDATE_THRESHOLD: Duration = Duration::from_secs(5);
+
+/// A trait that defines a receiver that receives events emitted from a [`TimedFuture`].
+///
+/// This trait is designed to allow any external consumer to hook into the lifecycle of a
+/// [`TimedFuture`].
+///
+/// An example usage could be to collect metrics, as implemented in [`RawMetrics`].
+pub trait TimedFutureReceiver {
+    /// Called before a poll is being performed by the [`TimedFuture`].
+    fn before_poll(&self) {}
+
+    /// Called after a poll was performed by the [`TimedFuture`].
+    fn after_poll(&self, _poll_duration: u64) {}
+
+    /// Called when the [`TimedFuture`] reaches the internal threshold for additional data.
+    fn on_threshold_reached(&self, _utilization: u8) {}
+}
```

Comment:
What's wrong with the existing `RawMetrics`? This seems awfully verbose and isn't easily future extensible.

Each caller will just have to basically copy the implementation of raw metrics, if you want to have one unified way of accessing metrics from multiple sources, they can be merged on access, then also without exposing the atomics in the `RawMetrics`.